lib/internal/modules/esm/translators.js

```diff
@@ -28,7 +28,16 @@ translators.set('esm', async (url) => {
   const source = `${await readFileAsync(new URL(url))}`;
   debug(`Translating StandardModule ${url}`);
   return {
-    module: new ModuleWrap(stripShebang(source), url),
+    module: new ModuleWrap(stripShebang(source), url, {
+      initializeImportMeta(meta) {
+        meta.url = url;
+      },
+      async importModuleDynamically(specifier) {
+        const loader = await require('internal/process/esm_loader')
+          .loaderPromise;
+        return loader.import(specifier, url);
+      },
+    }),
```

Comment:
It would be nice not to have to set these up for every module if we can just have a single general function that handles these as we do currently.... is it not possible to maintain the previous fallback mechanism while distinguishing contexts?

Reply:
fallback method leaves too much at risk imo. if a module wrap is nullptr for some reason I would rather a dynamic import request failed than went to the context loader.

Reply:
Alternatively since this object is the same for all ES modules, can we define this as a constant and pass the same object each time?

That does require receiving a second argument in importModuleDynamically to get the referrer though.

Reply:
I guess that won't really help since ModuleWrap picks these off anyway?

What I'm getting at is that any part of the module pipeline where we can reduce the amount of code running, we should, in the name of supporting fast loading of 1000s of modules. While we're setting this up now, it's worth thinking about, as refactoring of these types of interactions gets trickier as the code gets more complex.

It would just be nice to have a more streamlined approach than defining closures per-module, at least so far as there isn't any technical reason we can't.

Reply:
when I have time I'll set up some benchmarks for this

Reply:
Performance work will definitely be a great thing to start working on! But note that once you set up these closures, likely no one will ever change this to something simpler. The best time to write simple code is when you first write it, and conceptual simplicity itself is an important goal - performance aside too.

I'm not blocking anything on this, just commenting here.